File: elisp/manifest.cc
// Copyright 2020, 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "elisp/manifest.h"

#include <cstdlib>
#include <iostream>
#include <string>
#include <utility>
#include <vector>

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wpedantic"
#pragma GCC diagnostic ignored "-Wconversion"
#pragma GCC diagnostic ignored "-Wsign-conversion"
#pragma GCC diagnostic ignored "-Woverflow"
#include "absl/base/casts.h"
#include "absl/container/flat_hash_map.h"
#include "absl/random/random.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/types/optional.h"
#include "nlohmann/json.hpp"
#pragma GCC diagnostic pop

#include "elisp/algorithm.h"
#include "elisp/file.h"
#include "elisp/options.h"
#include "elisp/status.h"

namespace phst_rules_elisp {

absl::StatusOr<absl::optional<TempFile>> AddManifest(
    const Mode mode, std::vector<std::string>& args, absl::BitGen& random) {
  using Type = absl::optional<TempFile>;
  if (mode == Mode::kDirect) return absl::implicit_cast<Type>(absl::nullopt);
  ASSIGN_OR_RETURN(auto stream,
                   TempFile::Create(TempDir(), "manifest-*.json", random));
  args.push_back(absl::StrCat("--manifest=", stream.path()));
  args.push_back("--");
  return absl::implicit_cast<Type>(std::move(stream));
}

static void CheckRelative(const std::vector<std::string>& files) {
  for (const std::string& file : files) {
    if (IsAbsolute(file)) {
      std::clog << "filename " << file << " is absolute" << std::endl;
      std::abort();
    }
  }
}

absl::Status WriteManifest(const CommonOptions& opts,
                           std::vector<std::string> input_files,
                           const std::vector<std::string>& output_files,
                           TempFile& file) {
  CheckRelative(opts.load_path);
  CheckRelative(opts.load_files);
  const auto data_files = Sort(opts.data_files);
  CheckRelative(data_files);
  input_files.insert(input_files.end(),
                     opts.load_files.begin(), opts.load_files.end());
  input_files.insert(input_files.end(), data_files.begin(), data_files.end());
  const nlohmann::json json = {
      {"root", "RUNFILES_ROOT"},
      {"loadPath", opts.load_path},
      {"inputFiles", input_files},
      {"outputFiles", output_files},
      {"tags", Sort(opts.rule_tags)},
  };
  return file.Write(json.dump());
}

}  // namespace phst_rules_elisp

File: elisp/manifest.py
# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Functions to write Emacs Lisp wrapper manifests.

This is an internal helper library for the Emacs Lisp Bazel rules.  Don’t rely
on it in any way outside the rule implementation."""

import argparse
import contextlib
import json
import os
import os.path
import pathlib
import tempfile
from typing import Generator, IO, Iterable, List, Optional

@contextlib.contextmanager
def add(mode: str, args: List[str]) -> Generator[Optional[IO[str]], None, None]:
    """Create a temporary file for a manifest if needed.

If a file was created, appropriate arguments are added to args."""
    if mode not in ('direct', 'wrap'):
        raise ValueError(f'invalid mode {mode}')
    if mode == 'direct':
        yield None
    else:
        with tempfile.NamedTemporaryFile(
                mode='wt', encoding='utf-8',
                prefix='manifest-', suffix='.json') as file:
            args += ['--manifest=' + os.path.abspath(file.name), '--']
            yield file.file

def write(opts: argparse.Namespace, input_files: Iterable[pathlib.Path],
          output_files: Iterable[pathlib.Path], file: IO[str]) -> None:
    """Write manifest file contents to the given file object."""
    _check_relative(opts.load_directory)
    _check_relative(opts.load_file)
    data_files = sorted(opts.data_file)
    _check_relative(data_files)
    all_input_files = (tuple(input_files) +
                       tuple(opts.load_file) +
                       tuple(data_files))
    json_data = {
        'root': 'RUNFILES_ROOT',
        'loadPath': tuple(map(os.fspath, opts.load_directory)),
        'inputFiles': tuple(map(os.fspath, all_input_files)),
        'outputFiles': tuple(map(os.fspath, output_files)),
        'tags': sorted(opts.rule_tag),
    }
    json.dump(json_data, file)
    file.flush()

def _check_relative(files: Iterable[pathlib.PurePath]) -> None:
    for file in files:
        if file.is_absolute():
            raise ValueError(r'filename {file} is absolute')
